src/types.rs

```diff
@@ -0,0 +1,147 @@
+use pyo3::{
+    prelude::*,
+    types::{PyDict, PyList},
+    PyNativeType, ToBorrowedObject, ToPyObject,
+};
+
+pub trait PyArray<'py>: Into<PyObject> + Clone + 'py {
+    fn new<T, U>(py: Python<'py>, elements: impl IntoIterator<Item = T, IntoIter = U>) -> Self
+    where
+        T: ToPyObject,
+        U: ExactSizeIterator<Item = T>;
+
+    fn py(&self) -> Python<'py>;
+
+    fn len(&self) -> usize;
+
+    #[inline]
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn downcast(obj: &'py PyAny) -> PyResult<Self>;
+
+    fn is_instance(obj: &'py PyAny) -> PyResult<bool>;
+}
+
+pub trait PyDictionary<'py>: Into<PyObject> + Clone + 'py {
+    type PyIter: Iterator<Item = (PyObject, PyObject)>;
+
+    fn new(py: Python<'py>) -> Self;
```

Comment:
Similarly I think that we should have

```
fn pythonize_mapping(py: Python<'py>, elements: Vec<(PyObject, PyObject)>) -> PyObject
```

In this case, to handle the from-python direction, we need to build a `PyMapping` type for PyO3.

See https://github.com/PyO3/pyo3/blob/main/src/types/sequence.rs which is the implementation for `PySequence`. It should be possible to also build a `PyMapping` which uses [the mapping protocol methods](https://docs.python.org/3/c-api/mapping.html).

For now we can have this `PyMapping` struct as a private type inside `pythonize`, and then I'll push it upstream to PyO3 for the 0.15 release cycle.